uniref-rest/src/main/java/org/uniprot/api/uniref/queue/UniRefMessageListener.java

```diff
@@ -0,0 +1,115 @@
+package org.uniprot.api.uniref.queue;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.stream.Stream;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.core.MessageListener;
+import org.springframework.amqp.rabbit.core.RabbitTemplate;
+import org.springframework.amqp.support.converter.MessageConverter;
+import org.springframework.context.annotation.Profile;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Service;
+import org.uniprot.api.common.repository.stream.store.StoreRequest;
+import org.uniprot.api.rest.download.DownloadResultWriter;
+import org.uniprot.api.rest.download.model.DownloadJob;
+import org.uniprot.api.rest.download.model.JobStatus;
+import org.uniprot.api.rest.download.queue.AbstractMessageListener;
+import org.uniprot.api.rest.download.queue.AsyncDownloadQueueConfigProperties;
+import org.uniprot.api.rest.download.queue.DownloadConfigProperties;
+import org.uniprot.api.rest.download.queue.MessageListenerException;
+import org.uniprot.api.rest.download.repository.DownloadJobRepository;
+import org.uniprot.api.rest.output.UniProtMediaType;
+import org.uniprot.api.rest.output.context.FileType;
+import org.uniprot.api.rest.request.DownloadRequest;
+import org.uniprot.api.uniref.request.UniRefDownloadRequest;
+import org.uniprot.api.uniref.service.UniRefEntryLightService;
+
+/**
+ * @author tibrahim
+ * @created 16/08/2023
+ */
+@Profile({"live", "asyncDownload"})
+@Service("DownloadListener")
+@Slf4j
+public class UniRefMessageListener extends AbstractMessageListener implements MessageListener {
+    private static final String DATA_TYPE = "uniref";
+    private final MessageConverter converter;
+    private final UniRefEntryLightService service;
+    private final DownloadConfigProperties downloadConfigProperties;
+    private final DownloadResultWriter downloadResultWriter;
+
+    public UniRefMessageListener(
+            MessageConverter converter,
+            UniRefEntryLightService service,
+            DownloadConfigProperties downloadConfigProperties,
+            AsyncDownloadQueueConfigProperties asyncDownloadQueueConfigProperties,
+            DownloadJobRepository jobRepository,
+            DownloadResultWriter downloadResultWriter,
+            RabbitTemplate rabbitTemplate) {
+        super(
+                downloadConfigProperties,
+                asyncDownloadQueueConfigProperties,
+                jobRepository,
+                rabbitTemplate);
+        this.converter = converter;
+        this.service = service;
+        this.downloadConfigProperties = downloadConfigProperties;
+        this.downloadResultWriter = downloadResultWriter;
+    }
+
+    protected void processMessage(Message message, DownloadJob downloadJob) {
+        UniRefDownloadRequest request = (UniRefDownloadRequest) this.converter.fromMessage(message);
```

Comment:
processMessage is doing very similar thing in all the listener:
1. get request
2. get paths
3. check if seen before
4. write result based on condition
5. set status

This is the right candidate to use template method pattern.
Implement processMessage in AbstractListener and override special cases in the subclass:

```
void processMessage(Message message, DownloadJob downloadJob){
        DownloadRequest request = getRequest();
        Path idsFile = getIdsFilePath();
        Path resultsFile = getIdsFilePath();
        if (isJobSeenBefore(downloadJob, idsFile, resultFile)) {
            if (downloadJob.getStatus() == JobStatus.RUNNING) {
                log.warn("The job {} is running by other thread", jobId);
            } else {
                log.info("The job {} is already processed", jobId);
                updateDownloadJob(message, downloadJob, JobStatus.FINISHED);
            }
        } else {
            updateDownloadJob(message, downloadJob, JobStatus.RUNNING);
            writeResult(request, idsFile, jobId, contentType);
            updateDownloadJob(message, downloadJob, JobStatus.FINISHED, jobId);
        }
    }
```

Reply:
Yes, I agree. This improvement is out of scope for uniref task as this includes changes in idmapping, uniprotkb & uniref. We can create a separate ticket for this improvement and track it there.

Reply:
[TRM-30161](https://www.ebi.ac.uk/panda/jira/browse/TRM-30161)